crates/ark/src/lsp/signature_help.rs

```diff
@@ -336,13 +352,155 @@ fn is_within_call_parentheses(x: &Point, node: &Node) -> bool {
     x.is_after_or_equal(open.end_position()) && x.is_before_or_equal(close.start_position())
 }
 
+fn argument_label(name: String, value: SEXP) -> String {
+    // Specially handle `R_MissingArg`, which looks like a `SYMSXP`,
+    // but we don't want to add `=` to it. This is what we see when
+    // there is no default argument (and also for `...`).
+    if value == r_missing() {
+        return name;
+    }
+
+    name + " = " + obj_label(value).as_str()
+}
+
+fn obj_label(x: SEXP) -> String {
+    if r_is_object(x) {
+        // Just showing the class name for inlined complex objects
+        return s3_label(x);
+    }
+    if r_dim(x) != r_null() {
+        // In the odd case of an inlined array, just show type name
+        return type_label(x);
+    }
+
+    // Most cases are either:
+    // - `NULL`
+    // - Call
+    // - Symbol
+    // - Scalar objects that get inlined automatically
+    // However, it is possible to have more complex arguments inlined into the function
+    // signature, like vctrs does with `fn_inline_formals()`. We also handle most of
+    // those cases well.
+    match r_typeof(x) {
+        NILSXP => null_label(),
+        LANGSXP => call_label(x),
+        SYMSXP => sym_label(x),
+        LGLSXP => vec_label(x, lgl_to_pretty_string),
+        INTSXP => vec_label(x, int_to_pretty_string),
+        REALSXP => vec_label(x, dbl_to_pretty_string),
+        CPLXSXP => vec_label(x, cpl_to_pretty_string),
+        STRSXP => vec_label(x, chr_to_pretty_string),
+        VECSXP => list_label(),
+        _ => return type_label(x),
+    }
+}
+
+fn type_label(x: SEXP) -> String {
+    let out = r_type2char(r_typeof(x));
+    let out = String::from("<") + out.as_str() + ">";
+    out
+}
+
+fn s3_label(x: SEXP) -> String {
+    match r_s3_pretty_class(x) {
+        Ok(class) => class,
+        Err(err) => {
+            log::error!("{err:?}");
+            String::from("<???>")
+        },
+    }
+}
+
+fn null_label() -> String {
+    r_null_to_pretty_string()
+}
+
+fn list_label() -> String {
+    // TODO: Should we try and do better? I doubt an inlined list is that common.
```

Comment:
Probably best not to, especially if we need to consider whether someone inlined a data frame or model fit in the formals of their function 😱

Reply:
The most likely way this could happen is with code that generates functions from data.